OrbitGl/ThreadStateTrack.cpp

```diff
@@ -0,0 +1,168 @@
+    default:
+      UNREACHABLE();
+  }
+}
+
+static std::string GetThreadStateName(ThreadStateSliceInfo::ThreadState state) {
+  switch (state) {
+    case ThreadStateSliceInfo::kRunning:
+      return "Running";
+    case ThreadStateSliceInfo::kRunnable:
+      return "Runnable";
+    case ThreadStateSliceInfo::kInterruptibleSleep:
+      return "Interruptible sleep";
+    case ThreadStateSliceInfo::kUninterruptibleSleep:
+      return "Uninterruptible sleep";
+    case ThreadStateSliceInfo::kStopped:
+      return "Stopped";
+    case ThreadStateSliceInfo::kTraced:
+      return "Traced";
+    case ThreadStateSliceInfo::kDead:
+      return "Dead";
+    case ThreadStateSliceInfo::kZombie:
+      return "Zombie";
+    case ThreadStateSliceInfo::kParked:
+      return "Parked";
+    case ThreadStateSliceInfo::kIdle:
+      return "Idle";
+    default:
+      UNREACHABLE();
+  }
+}
+
+std::string ThreadStateTrack::GetThreadStateSliceTooltip(PickingId id) const {
+  auto user_data = time_graph_->GetBatcher().GetUserData(id);
+  if (user_data == nullptr || user_data->custom_data_ == nullptr) {
+    return "";
+  }
+
+  const auto* thread_state_slice =
+      static_cast<const ThreadStateSliceInfo*>(user_data->custom_data_);
+  return absl::StrFormat(
+      "<b>%s</b><br/>"
+      "<i>Thread state</i><br/>",
+      GetThreadStateName(thread_state_slice->thread_state()));
+}
+
+void ThreadStateTrack::UpdatePrimitives(uint64_t min_tick, uint64_t max_tick,
+                                        PickingMode /*picking_mode*/) {
+  Batcher* batcher = &time_graph_->GetBatcher();
+
+  const auto time_window_ns = static_cast<uint64_t>(1000 * time_graph_->GetTimeWindowUs());
+  const uint64_t pixel_delta_ns = time_window_ns / time_graph_->GetCanvas()->GetWidth();
+  const uint64_t min_time_graph_ns = time_graph_->GetTickFromUs(time_graph_->GetMinTimeUs());
+
+  uint64_t ignore_until_ns = 0;
+
+  GOrbitApp->GetCaptureData().ForEachThreadStateSliceIntersectingTimeRange(
+      thread_id_, min_tick, max_tick, [&](const ThreadStateSliceInfo& slice) {
+        if (slice.end_timestamp_ns() <= ignore_until_ns) {
+          // Reduce overdraw by not drawing slices whose entire width would only draw over a
+          // previous slice. Similar to TimerTrack::UpdatePrimitives.
+          return;
+        }
+
+        const float x0 =
+            std::max(pos_[0], time_graph_->GetWorldFromTick(slice.begin_timestamp_ns()));
+        const float x1 =
+            std::min(pos_[0] + size_[0], time_graph_->GetWorldFromTick(slice.end_timestamp_ns()));
+        const float width = x1 - x0;
+
+        const Vec2 pos{x0, pos_[1]};
+        const Vec2 size{width, -size_[1]};
+
+        const Color color = GetThreadStateColor(slice.thread_state());
+
+        auto user_data = std::make_unique<PickingUserData>(
+            nullptr, [&](PickingId id) { return GetThreadStateSliceTooltip(id); });
+        user_data->custom_data_ = &slice;
+
+        if (width > 1.0f) {
```

Comment:
I'll look into it.

Reply:
Ok, in the meantime, I'm a bit confused with world coordinates and pixels. For example, `GlCanvas::GetWorldWidth` returns `world_width_` and `GlCanvas::GetWidth` returns `width_` (which should be the width in pixels), but in `GlCanvas::Prepare2DViewport` we have `world_width_ = static_cast<float>(width_)`. So is there a deeper difference other than the type?

Reply:
So this is done, please have a look.

But below I use `Box box(pos, {1.0f, size[1]}, GlCanvas::kZValueEvent);`.
Is that `1.0f` also wrong? How do I find the width of a pixel in world coordinates?
Should this instead be `time_graph_->GetCanvas()->GetWorldWidth() / time_graph_->GetCanvas()->GetWidth()`? And as I said above, isn't this 1.0f anyway?

Reply:
Yes, in this case you'd also need to use `pixel_delta_ns` instead of `1.0f`, because you specify the box coordinates in world space, but you want them to be 1px in screen space.

But I have to admit that I am now also confused about the line you pointed out in Prepare2DViewport, because world space size and pixel width should not be the same, and I was under the assumption that the glOrtho2D projection matrix takes care of horizontal zoom, which apparently it doesn't? Which makes your arguments absolutely valid, nevertheless it's a good idea to assume at this point that the coordinate systems are different and to convert accordingly.

Reply:
I used `canvas->GetWorldWidth() / static_cast<float>(canvas->GetWidth())` as I had said above. I'm pretty convinced that world space is not in nanoseconds and that using `pixel_delta_ns` causes very very long slices (that then don't like like so because the slices that come after are drawn on top). To me it just seems it's pixels, but float, but maybe there's some scaling. Anyway, let's hear from Pierric.

Reply:
It's true that currently the "world" width matches the width in pixels, but this wasn't always the case, and this could change in the future. As Florian mentioned, we could implement zooming and panning by only modifying the viewport, but to minimize floating point accuracy issues, the timers are always repositioned and rescaled relative to the current time window. That being said, we should not assume that the screen and world coordinate systems match in any way even if in the current implementation it "kinda" does.

Reply:
Ok, but this should be made much clearer, in particular what world coordinates represent and what their reference is. Because at the moment, they don't "kinda" match screen coordinates, they are exactly the screen coordinates.